src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxResolvingClassLoader.java

```diff
@@ -93,12 +93,18 @@
         });
     }
 
-    private static <T> LoadingCache<ClassLoader, Cache<String, T>> makeParentCache(boolean weakValues) {
-        Caffeine<Object, Object> builder = Caffeine.newBuilder().weakKeys();
-        if (weakValues) {
-            builder = builder.weakValues();
+    private static <T> LoadingCache<ClassLoader, Cache<String, T>> makeParentCache(boolean weakValuesInnerCache) {
+        // The outer cache has weak keys, so that we do not leak class loaders, but strong values, because the
+        // inner caches are only referenced by the outer cache internally.
+        Caffeine<Object, Object> outerBuilder = Caffeine.newBuilder().weakKeys();
+        // The inner cache has strong keys, since they are just strings, and expires entries 15 minutes after they are
+        // added to the cache (TODO: should we use expireAfterAccess instead?). The values for the inner cache may
+        // be weak if needed, for example parentClassCache uses weak values to avoid leaking classes and their loaders.
+        Caffeine<Object, Object> innerBuilder = Caffeine.newBuilder().expireAfterWrite(Duration.ofMinutes(15));
+        if (weakValuesInnerCache) {
+            innerBuilder.weakValues();
         }
 
-        return builder.build(parentLoader -> Caffeine.newBuilder()./* allow new plugins to be used, and clean up memory */expireAfterWrite(Duration.ofMinutes(15)).build());
```

Comment:
Might want to retain a mention in comments that it is important to expire entries _somehow_ in order to let new plugins be used. Ideally this would be done with a listener.

Reply:
Ah ok, it wasn't totally clear to me what that comment was referring to. I will add it back.